src/ydata_synthetic/synthesizers/timeseries/doppelganger/model.py

```diff
@@ -0,0 +1,229 @@
+        d_regularizer = reduce_mean((ddx - 1.0) ** 2)
+        return d_regularizer
+
+    def train(self, data, train_args):
+        super().train(data=data, num_cols=None, cat_cols=None)
+        self.define_gan()
+
+def get_noise_sample(batch_size: int, noise_dim: int):
+    return normal([batch_size, noise_dim])
+
+class RealMGenerator(Model):
+    """'Real' metadata generator.
+    Produces the metadata associated with each sequence."""
+    def __init__(self, meta_dim: int, noise_dim: int, dim: int=100, activation_info: Optional[NamedTuple]=None,
+                    tau: Optional[float] = None, batch_size: int=64):
+        """Arguments:
+            meta_dim(int): The cardinality of the produced metadata
+            noise_dim(int): The cardinality of the noise array internally consumed by the model
+            dim(int): The number of units used in the model layers
+            activation_info(Optional[NamedTuple]): Specifies the type of activations to be used for the output
+            tau(Optional[float]): Tau parameter of the Gumbel-Softmax (if categorical columns were specified)
+            batch_size(int): Number of sequences passed to the model in train time"""
+        super().__init__()
+
+        self.batch_size = batch_size
+        self.noise_dim = noise_dim
+        self.meta_dim = meta_dim
+        self.dim = dim
+        self.activation_info = activation_info
+        self.tau = tau
+
+    def build(self, _):
+        self.model = Sequential(name='Real Metadata Generator')
+        self.model.add(Dense(self.dim))
+        self.model.add(Dense(self.dim))
+        self.model.add(Dense(self.meta_dim))
+        if self.activation_info:
+            self.model.add(GumbelSoftmaxActivation(self.activation_info, tau=self.tau))
+
+    def call(self, _, training=False) -> Tensor:
+        noise = get_noise_sample(self.batch_size, self.noise_dim)
+        return self.model(noise)
+
+class FakeMGenerator(Model):
+    """'Fake' metadata generator.
+    Produces min and max of each timeseries."""
+    def __init__(self, noise_dim: int, ts_dim: Tuple[int], dim: int=100, batch_size: int=64):
+        """Arguments:
+            noise_dim(int): The cardinality of the noise array internally consumed by the model
+            dim(int): The number of units used in the model layers
+            ts_dim(Tuple[int]): Tuple with dimensions of the synthesized time series as: (seq_length, n_series)
+            batch_size(int): Number of sequences passed to the model in train time"""
+        super().__init__()
+
+        self.batch_size = batch_size
+        self.noise_dim = noise_dim
+        self.dim = dim
+        _, self.n_ts = ts_dim
+
+    def build(self, _):
+        self.model = Sequential(name='Fake Metadata Generator')
+        self.model.add(Dense(self.dim))
+        self.model.add(Dense(self.dim))
+        self.model.add(Dense(self.n_ts*2))
+
+    def call(self, metadata: Tensor, training=False) -> Tensor:
```

Comment:
For all the other GANs such as timeGAN this logic was not set at the network level, but rather within the main class of the algorithm. 

I suggest to follow the same logic if possible to keep things simple